### DISCUSSION TITLE
Sudo with gum spin

### DISCUSSION THREAD

**Discussion by riasat-sheikh** [charmbracelet/gum#104] [Q&A]
How to implement a sudo command with gum spin?

For example,

`gum spin --show-output --spinner meter --title "Updating LaTeX utility..." -- sudo tlmgr update --self`

**Comment by dakotak** [comment]
One solution that may work is to check if sudo will require a password. It's pretty common for sudo to to not require a password for some time (configured used the `timestamp_timeout` variable in the sudoers file) if it has already authenticated the user during the session.

```shell
if sudo -n true &> /dev/null; then
  gum ... -- sudo your_command
else
  sudo your_command
fi
```

If you want to increase the chances of this method working you can do something like `sudo --validate` before the if condition to have sudo prompt and potentially cache the escalation so that on the next sudo usage credentials may not be required. But keep in mind that sudo can be configured to require a password every time or even other authentication mechanisms  so make sure your script can handle running the command without `gum spin` to allow sudo to be interactive.

You may also want to check out `sudo --askpass`. There may also be a way to to check if sudo will allow the user to cache credentials.

From the man page:
> The policy may require that users authenticate themselves with a password or another authentication mechanism. If authentication is required, sudo will exit if the user's password is not entered within a configurable time limit.  This limit is policy-specific; the default password prompt timeout for the sudoers security policy is 5 minutes.
>
> Security policies may support credential caching to allow the user to run sudo again for a period of time without requiring authentication.  By default, the sudoers policy caches credentials on a per-terminal basis for 5 minutes.  See the timestamp_type and timestamp_timeout options in sudoers(5) for more information.  By running sudo with the -v option, a user can update  the  cached credentials without running a command.

**Comment by beltagyy** [comment]
gum spin forks your command into a background process and hides stdout until completion. This breaks any interactive prompt -- including sudo's password request -- because stdin/stderr of the forked process is captured rather than connected to your terminal.\n\nWorkaround options:\n\n1. Cache sudo first:\nsudo -v\ngum spin --show-output --spinner meter --title "Updating LaTeX..." -- sudo tlmgr update --self\n\n2. Use a PTY wrapper (if you need a password prompt):\ngum spin --show-output --spinner meter --title "Updating LaTeX..." -- script -qec "sudo tlmgr update --self" /dev/null\n\n3. Separate phases:\ngum confirm "Continue with LaTeX update?" && gum spin ...
